cognite_synthetic_tags/tag_resolver.py

```diff
@@ -0,0 +1,333 @@
+            # some operands are instance of `pd.Series`, we need to
+            # apply the operation element-wise
+            values_series: List[pd.Series] = operands  # keeping mypy happy
+            result = pd.Series(
+                (
+                    operation(*[series[i] for series in values_series])
+                    for i in series_index
+                ),
+                index=series_index,
+            )
+        else:
+            # all operands are just numbers, apply the operation directly
+            result = operation(*operands)
+
+        return result
+
+    @staticmethod
+    def _extract_literals_from_specs(
+        specs: TagSpecsT,
+    ) -> Tuple[TagSpecsT, TagResolverContextT]:
+        """
+        Split specs dict in two: one with the actual specs and another with
+        only literal values.
+        """
+        new_specs: TagSpecsT = {}
+        literals: TagResolverContextT = {}
+        for key, value in specs.items():
+            if isinstance(value, Tag):
+                new_specs[key] = value
+            else:
+                literals[key] = value
+        return new_specs, literals
+
+    @staticmethod
+    def _make_series(
+        data: List[TagValueT],
+    ) -> Tuple[List[TagValueT], Optional[pd.Index]]:
+        """
+        Take a list of values, and if any of the items is a `pd.Series`
+        instance, change other values into `pd.Series` as well.
+
+        Returns a 2-tuple:
+        - list of values, either unchanged or with each item a `pd.Series`,
+        - an index for the series or None. All series have the same index.
+        """
+        series: List[pd.Series] = [
+            val for val in data if isinstance(val, pd.Series)
+        ]
+        if not series:
+            return data, None
+
+        # check that all series have the same indexes:
+        #   (all are returned from the same CDF API call, so they should)
+        index = series[0].index
+        for single_series in series:
+            assert all(
+                single_series.index == index
+            ), "Series need to have the same index."
+        # convert any non-series items to series:
+        #   (repeat the item for every index)
+        all_series: List[pd.Series] = [
+            val
+            if isinstance(val, pd.Series)
+            else pd.Series([val] * len(index), index=index)
+            for val in data
+        ]
+
+        return all_series, index
+
+    @staticmethod
+    def _make_series_in_dict(
+        data: TagResolverContextT,
+        index: pd.Index,
+    ) -> TagResolverContextT:
+        """
+        Make sure that all values in the dict are `pd.Series` instances if the
+        index is a `pd.Index`.
+        """
+        if index is not None:
+            data["__dummy_series__"] = pd.Series({}, index=index)
+            # TODO ^^ Yeah, it's a magic string...  ¯\_(ツ)_/¯
+            series_or_values, _ = TagResolver._make_series(list(data.values()))
```

Comment:
Might also need to change type annotation to `Iterable` or similar
```suggestion
            series_or_values, _ = TagResolver._make_series(data.values())
```